wrappers/services.go

```diff
@@ -244,139 +242,97 @@ func StartServices(apps []*snap.AppInfo, disabledSvcs *DisabledServices, opts *S
 	}
 
 	systemSysd := systemd.New(systemd.SystemMode, inter)
-	userSysd := systemd.New(systemd.GlobalUserMode, inter)
+	userGlobalSysd := systemd.New(systemd.GlobalUserMode, inter)
 	cli, err := newUserServiceClientNames(opts.Users, inter)
 	if err != nil {
 		return err
 	}
 
-	var toEnableSystem []string
-	var toEnableUser []string
-	systemServices := make([]string, 0, len(apps))
-	userServices := make([]string, 0, len(apps))
-	servicesStarted := false
+	// When starting and enabling services, act on the activated units instead of
+	// the services activated by those units. And since 'static' service units does
+	// not need to be enabled, we can save that.
+	const includeActivatedServices = false
 
+	filteredApps := filterServicesForStart(apps, disabledSvcs, opts.Scope)
+	systemServices, userServices := categorizeServices(filteredApps, includeActivatedServices)
+	var undoStart bool
 	defer func() {
 		if err == nil {
 			return
 		}
-		// apps could have been sorted according to their startup
-		// ordering, stop them in reverse order
-		if servicesStarted {
-			for i := len(apps) - 1; i >= 0; i-- {
-				app := apps[i]
+
+		// Undo logic for user services is handled by user session agent,
```

Comment:
It seems that the undo logic by user session agent only disabled enabled services, but never stops them? I might be missing context here.

https://github.com/snapcore/snapd/blob/a77998cf56fc6834c9c58703d9de4e5b2a4ad796/usersession/agent/rest_api.go#L107-L125

Reply:
It does stop any service it managed to start here:
https://github.com/snapcore/snapd/blob/a77998cf56fc6834c9c58703d9de4e5b2a4ad796/usersession/agent/rest_api.go#L140-L147

Reply:
my bad, It seems I missed that bit somehow